File: Uni_MS/uni_ms/src/main/java/com/badrulamin/University_Management/service/AdmissionMeritListService.java
package com.badrulamin.University_Management.service;

import com.badrulamin.University_Management.entity.*;
import com.badrulamin.University_Management.exception.BusinessException;
import com.badrulamin.University_Management.exception.ResourceNotFoundException;
import com.badrulamin.University_Management.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class AdmissionMeritListService {

    private final AdmissionMeritListRepository meritListRepository;
    private final AdmissionMeritListEntryRepository entryRepository;
    private final AdmissionTestAttemptRepository attemptRepository;
    private final AdmissionTestResultRepository testResultRepository;
    private final AdmissionTestRepository testRepository;
    private final PreAdmissionRegistrationRepository registrationRepository;
    private final NotificationHelper notificationHelper;

    public Page<AdmissionMeritList> findAll(Pageable pageable) {
        return meritListRepository.findAll(pageable);
    }

    public Page<AdmissionMeritList> findByFilters(String search, String status, Long sessionId,
            Long facultyId, Long departmentId, Long programId, Long testId, Pageable pageable) {
        return meritListRepository.findByFilters(search, status, sessionId, facultyId,
                departmentId, programId, testId, pageable);
    }

    public AdmissionMeritList findById(Long id) {
        return meritListRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("MeritList", "id", id));
    }

    @Transactional
    public AdmissionMeritList save(AdmissionMeritList meritList) {
        return meritListRepository.save(meritList);
    }

    @Transactional
    public AdmissionMeritList update(Long id, AdmissionMeritList meritList) {
        findById(id);
        meritList.setId(id);
        return meritListRepository.save(meritList);
    }

    @Transactional
    public void delete(Long id) {
        AdmissionMeritList list = findById(id);
        if ("PUBLISHED".equals(list.getStatus())) {
            throw new BusinessException("Cannot delete a published merit list. Unpublish it first.");
        }
        entryRepository.deleteAll(entryRepository.findByMeritList_IdOrderByRankAsc(id));
        meritListRepository.deleteById(id);
    }

    public long countByStatus(String status) {
        return meritListRepository.countByStatus(status);
    }

    public Map<String, Object> getStats() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("total", meritListRepository.count());
        stats.put("draft", meritListRepository.countByStatus("DRAFT"));
        stats.put("published", meritListRepository.countByStatus("PUBLISHED"));
        stats.put("archived", meritListRepository.countByStatus("ARCHIVED"));
        return stats;
    }

    @Transactional
    public AdmissionMeritList generateMeritList(Long testId, String listName, Integer totalSeats,
            String academicYear, Long facultyId, Long departmentId, Long programId, String shift) {
        AdmissionTest test = testRepository.findById(testId)
                .orElseThrow(() -> new ResourceNotFoundException("AdmissionTest", "id", testId));

        List<AdmissionTestAttempt> gradedAttempts = attemptRepository
                .findByTest_IdAndStatusOrderByScoreDesc(testId, "GRADED");

        AdmissionMeritList meritList = AdmissionMeritList.builder()
                .name(listName != null ? listName : test.getName() + " - Merit List")
                .description("Auto-generated merit list for " + test.getName())
                .academicYear(academicYear != null ? academicYear : test.getAcademicYear())
                .session(test.getSession())
                .faculty(facultyId != null ? getFaculty(facultyId) : test.getFaculty())
                .department(departmentId != null ? getDepartment(departmentId) : test.getDepartment())
                .program(programId != null ? getProgram(programId) : test.getProgram())
                .shift(shift != null ? shift : test.getShift())
                .test(test)
                .totalSeats(totalSeats)
                .status("DRAFT")
                .build();
        meritList = meritListRepository.save(meritList);

        List<AdmissionMeritListEntry> entries = new ArrayList<>();
        int rank = 1;

        for (AdmissionTestAttempt attempt : gradedAttempts) {
            if (totalSeats != null && rank > totalSeats) break;

            PreAdmissionRegistration reg = attempt.getRegistration();

            Double academicScore = calculateAcademicScore(reg);
            Double totalWeighted = calculateTotalWeightedScore(attempt.getScore(), academicScore, test.getTotalMarks());

            AdmissionMeritListEntry entry = AdmissionMeritListEntry.builder()
                    .meritList(meritList)
                    .registration(reg)
                    .rank(rank)
                    .rollNumber("R" + String.format("%04d", rank))
                    .applicationNumber(reg.getRegistrationNumber())
                    .applicantName(reg.getFirstName() + " " + reg.getLastName())
                    .facultyName(test.getFaculty() != null ? test.getFaculty().getName() : "N/A")
                    .departmentName(test.getDepartment() != null ? test.getDepartment().getName() : "N/A")
                    .programName(test.getProgram() != null ? test.getProgram().getName() : "N/A")
                    .shift(test.getShift())
                    .testMarks(attempt.getScore())
                    .testMaxMarks(attempt.getMaxScore())
                    .score(attempt.getScore())
                    .academicScore(academicScore)
                    .totalWeightedScore(totalWeighted)
                    .sscGpa(reg.getSscGpa())
                    .hscGpa(reg.getHscGpa())
                    .quotaType("GENERAL")
                    .status(rank <= (totalSeats != null ? totalSeats : Integer.MAX_VALUE) ? "SELECTED" : "WAITING")
                    .submittedAt(attempt.getSubmittedAt())
                    .build();
            entries.add(entryRepository.save(entry));
            rank++;
        }

        meritList.setTotalApplicants(gradedAttempts.size());
        meritList.setSelectedCount((int) entries.stream().filter(e -> "SELECTED".equals(e.getStatus())).count());
        meritList.setWaitingCount((int) entries.stream().filter(e -> "WAITING".equals(e.getStatus())).count());

        if (!entries.isEmpty()) {
            meritList.setCutoffScore(entries.stream()
                    .mapToDouble(e -> e.getTotalWeightedScore() != null ? e.getTotalWeightedScore() : 0.0)
                    .min().orElse(0.0));
        }

        return meritListRepository.save(meritList);
    }

    @Transactional
    public AdmissionMeritList publish(Long id, String publishedBy) {
        AdmissionMeritList meritList = findById(id);
        if ("PUBLISHED".equals(meritList.getStatus())) {
            throw new BusinessException("Merit list is already published");
        }
        meritList.setStatus("PUBLISHED");
        meritList.setPublishedAt(LocalDateTime.now());
        meritList.setPublishedBy(publishedBy);
        return meritListRepository.save(meritList);
    }

    @Transactional
    public AdmissionMeritList unpublish(Long id) {
        AdmissionMeritList meritList = findById(id);
        if (!"PUBLISHED".equals(meritList.getStatus())) {
            throw new BusinessException("Merit list is not published");
        }
        meritList.setStatus("DRAFT");
        meritList.setPublishedAt(null);
        meritList.setPublishedBy(null);
        return meritListRepository.save(meritList);
    }

    @Transactional
    public AdmissionMeritList archive(Long id) {
        AdmissionMeritList meritList = findById(id);
        meritList.setStatus("ARCHIVED");
        return meritListRepository.save(meritList);
    }

    private Double calculateAcademicScore(PreAdmissionRegistration reg) {
        double ssc = reg.getSscGpa() != null ? reg.getSscGpa() : 0.0;
        double hsc = reg.getHscGpa() != null ? reg.getHscGpa() : 0.0;
        return (ssc + hsc) / 2.0;
    }

    private Double calculateTotalWeightedScore(Double testScore, Double academicScore, Integer totalMarks) {
        double testWeight = totalMarks != null && totalMarks > 0 ? (testScore / totalMarks) * 70 : 0;
        double academicWeight = academicScore * 30;
        return testWeight + academicWeight;
    }

    private Faculty getFaculty(Long id) {
        Faculty f = new Faculty();
        f.setId(id);
        return f;
    }

    private Department getDepartment(Long id) {
        Department d = new Department();
        d.setId(id);
        return d;
    }

    private Program getProgram(Long id) {
        Program p = new Program();
        p.setId(id);
        return p;
    }
}

File: Uni_MS/angular-ui/src/app/services/admission-merit-list.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AdmissionMeritList } from '../models/admission-merit-list';
import { AdmissionMeritListEntry } from '../models/admission-merit-list-entry';
import { environment } from '../../environments/environment';
import { PagedResponse, PageParams, DEFAULT_PAGE_PARAMS } from '../models/paged-response';

@Injectable({ providedIn: 'root' })
export class AdmissionMeritListService {
  private apiUrl = `${environment.apiUrl}/admission-merit-lists`;
  constructor(private http: HttpClient) {}

  findAll(params: PageParams = DEFAULT_PAGE_PARAMS, filters: any = {}): Observable<PagedResponse<AdmissionMeritList>> {
    let httpParams = new HttpParams()
      .set('page', params.page.toString())
      .set('size', params.size.toString())
      .set('sortBy', params.sortBy)
      .set('sortDir', params.sortDir);
    if (filters.search) httpParams = httpParams.set('search', filters.search);
    if (filters.status) httpParams = httpParams.set('status', filters.status);
    if (filters.sessionId) httpParams = httpParams.set('sessionId', filters.sessionId.toString());
    if (filters.facultyId) httpParams = httpParams.set('facultyId', filters.facultyId.toString());
    if (filters.departmentId) httpParams = httpParams.set('departmentId', filters.departmentId.toString());
    if (filters.programId) httpParams = httpParams.set('programId', filters.programId.toString());
    if (filters.testId) httpParams = httpParams.set('testId', filters.testId.toString());
    return this.http.get<PagedResponse<AdmissionMeritList>>(this.apiUrl, { params: httpParams });
  }

  findById(id: number): Observable<AdmissionMeritList> {
    return this.http.get<AdmissionMeritList>(`${this.apiUrl}/${id}`);
  }

  save(data: AdmissionMeritList): Observable<AdmissionMeritList> {
    return this.http.post<AdmissionMeritList>(this.apiUrl, data);
  }

  update(id: number, data: AdmissionMeritList): Observable<AdmissionMeritList> {
    return this.http.put<AdmissionMeritList>(`${this.apiUrl}/${id}`, data);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  generate(testId: number, listName?: string, totalSeats?: number, academicYear?: string,
           facultyId?: number, departmentId?: number, programId?: number, shift?: string): Observable<AdmissionMeritList> {
    let params = new HttpParams().set('testId', testId.toString());
    if (listName) params = params.set('listName', listName);
    if (totalSeats) params = params.set('totalSeats', totalSeats.toString());
    if (academicYear) params = params.set('academicYear', academicYear);
    if (facultyId) params = params.set('facultyId', facultyId.toString());
    if (departmentId) params = params.set('departmentId', departmentId.toString());
    if (programId) params = params.set('programId', programId.toString());
    if (shift) params = params.set('shift', shift);
    return this.http.post<AdmissionMeritList>(`${this.apiUrl}/generate`, {}, { params });
  }

  publish(id: number): Observable<AdmissionMeritList> {
    return this.http.put<AdmissionMeritList>(`${this.apiUrl}/${id}/publish`, {});
  }

  unpublish(id: number): Observable<AdmissionMeritList> {
    return this.http.put<AdmissionMeritList>(`${this.apiUrl}/${id}/unpublish`, {});
  }

  archive(id: number): Observable<AdmissionMeritList> {
    return this.http.put<AdmissionMeritList>(`${this.apiUrl}/${id}/archive`, {});
  }

  getStats(): Observable<any> {
    return this.http.get(`${this.apiUrl}/stats`);
  }

  getEntries(meritListId: number, params: PageParams = DEFAULT_PAGE_PARAMS, filters: any = {}): Observable<PagedResponse<AdmissionMeritListEntry>> {
    let httpParams = new HttpParams()
      .set('page', params.page.toString())
      .set('size', params.size.toString())
      .set('sortBy', params.sortBy)
      .set('sortDir', params.sortDir);
    if (filters.search) httpParams = httpParams.set('search', filters.search);
    if (filters.status) httpParams = httpParams.set('status', filters.status);
    if (filters.quotaType) httpParams = httpParams.set('quotaType', filters.quotaType);
    return this.http.get<PagedResponse<AdmissionMeritListEntry>>(`${this.apiUrl}/${meritListId}/entries`, { params: httpParams });
  }

  getAllEntries(meritListId: number): Observable<AdmissionMeritListEntry[]> {
    return this.http.get<AdmissionMeritListEntry[]>(`${this.apiUrl}/${meritListId}/entries/all`);
  }

  updateEntryStatus(entryId: number, status: string): Observable<AdmissionMeritListEntry> {
    return this.http.put<AdmissionMeritListEntry>(`${this.apiUrl}/entries/${entryId}/status`, null, { params: { status } });
  }

  deleteEntry(entryId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/entries/${entryId}`);
  }
}
